File: sort.py
class Sort:
    def quick(self, items):
        if len(items) < 2:
            return items
        pivot = items[0]
        smaller = [item for item in items if item < pivot]
        bigger = [item for item in items if item > pivot]
        equal = [item for item in items if item == pivot]
        return self.quick(smaller) + equal + self.quick(bigger)

    def bubble(self, items):
        length = len(items)
        for passover in range(length):
            for index in range(length - 1):
                if items[index] > items[index + 1]:
                    items[index], items[index + 1] = items[index + 1], items[index]
        return items

    def select(self, items):
        length = len(items)
        for passes in range(length):
            minimum = passes
            for i in range(passes, length):
                if items[i] < items[minimum]:
                    minimum = i
            if minimum != passes:
                items[passes], items[minimum] = items[minimum], items[passes]
        return items

    def insert(self, items):
        length = len(items)
        for i in range(1, length):
            index = i - 1
            temporary = items[i]
            while index >= 0 and items[index] > temporary:
                items[index + 1] = items[index]
                index -= 1
            items[index + 1] = temporary
        return items

    def simple(self, items):
        length = len(items)
        for i in range(length):
            for j in range(length):
                if items[i] < items[j]:
                    items[i], items[j] = items[j], items[i]
        return items

    def merge_sort(self, items):
        if len(items) <= 1:
            return items
        middle = len(items) // 2
        left = self.merge_sort(items[:middle])
        right = self.merge_sort(items[middle:])
        return self.merge(left, right)

    def merge(self, left, right):
        merged = []
        while len(left) > 0 and len(right) > 0:
            if left[0] <= right[0]:
                merged.append(left[0])
                left = left[1:]
            else:
                merged.append(right[0])
                right = right[1:]
        merged.extend(left)
        merged.extend(right)
        return merged


arr = [0, 43, 3, 2, 3, 4, 6]
sort = Sort()

quick_sort = sort.quick(arr)
select_sort = sort.select(arr)
insert_sort = sort.insert(arr)
bubble_sort = sort.bubble(arr)
simple_sort = sort.simple(arr)
merge_sort = sort.merge_sort(arr)

print("quick sort:", quick_sort)
print("select sort:", select_sort)
print("insert sort:", insert_sort)
print("bubble sort:", bubble_sort)
print("simple sort:", simple_sort)
print("merge sort:", merge_sort)

File: sort.go
package main

import (
	"fmt"
	"sync"
)

type Sort struct{}

func (s *Sort) Quick(items []int) []int {
	list := make([]int, len(items))
	copy(list, items)
	if len(list) < 2 {
		return list
	}
	pivot := list[0]
	smaller := make([]int, 0)
	bigger := make([]int, 0)
	for _, item := range list[1:] {
		if item < pivot {
			smaller = append(smaller, item)
		} else {
			bigger = append(bigger, item)
		}
	}
	return append(append(s.Quick(smaller), pivot), s.Quick(bigger)...)
}

func (s *Sort) Bubble(items []int) []int {
	length := len(items)
	for passover := 0; passover < length; passover++ {
		for index := 0; index < length-passover-1; index++ {
			if items[index] > items[index+1] {
				items[index], items[index+1] = items[index+1], items[index]
			}
		}
	}
	return items
}

func (s *Sort) Select(items []int) []int {
	length := len(items)
	for passes := 0; passes < length; passes++ {
		min := passes
		for i := passes; i < length; i++ {
			if items[i] < items[min] {
				min = i
			}
		}
		if min != passes {
			items[passes], items[min] = items[min], items[passes]
		}
	}
	return items
}

func (s *Sort) Insert(items []int) []int {
	length := len(items)
	for i := 1; i < length; i++ {
		index := i - 1
		temporary := items[i]
		for index >= 0 && items[index] > temporary {
			items[index+1] = items[index]
			index--
		}
		items[index+1] = temporary
	}
	return items
}

func (s *Sort) Simple(items []int) []int {
	length := len(items)
	for i := 0; i < length; i++ {
		for j := 0; j < length; j++ {
			if items[i] < items[j] {
				items[i], items[j] = items[j], items[i]
			}
		}
	}
	return items
}

func (s *Sort) MergeSort(items []int) []int {
	length := len(items)
	if length <= 1 {
		return items
	}
	middle := length / 2
	left := s.MergeSort(items[:middle])
	right := s.MergeSort(items[middle:])
	return s.merge(left, right)
}

func (s *Sort) merge(left, right []int) []int {
	merged := make([]int, 0)
	leftIndex := 0
	rightIndex := 0
	for leftIndex < len(left) && rightIndex < len(right) {
		if left[leftIndex] < right[rightIndex] {
			merged = append(merged, left[leftIndex])
			leftIndex++
		} else {
			merged = append(merged, right[rightIndex])
			rightIndex++
		}
	}
	for leftIndex < len(left) {
		merged = append(merged, left[leftIndex])
		leftIndex++
	}
	for rightIndex < len(right) {
		merged = append(merged, right[rightIndex])
		rightIndex++
	}
	return merged
}

func main() {
	arr := []int{0, 43, 3, 2, 3, 4}
	sort := Sort{}

	wg := sync.WaitGroup{}
	wg.Add(6)

	go func() {
		defer wg.Done()
		quickSort := sort.Quick(arr)
		fmt.Println("quick sort:", quickSort)
	}()

	go func() {
		defer wg.Done()
		selectSort := sort.Select(arr)
		fmt.Println("select sort:", selectSort)
	}()

	go func() {
		defer wg.Done()
		insertSort := sort.Insert(arr)
		fmt.Println("insert sort:", insertSort)
	}()

	go func() {
		defer wg.Done()
		bubbleSort := sort.Bubble(arr)
		fmt.Println("bubble sort:", bubbleSort)
	}()

	go func() {
		defer wg.Done()
		simpleSort := sort.Simple(arr)
		fmt.Println("simple sort:", simpleSort)
	}()

	go func() {
		defer wg.Done()
		mergeSort := sort.MergeSort(arr)
		fmt.Println("merge sort:", mergeSort)
	}()

	wg.Wait()
}
